texts/functions.dx

```diff
@@ -0,0 +1,242 @@
+of the argument, and then the `Ix` constraint is satisfied by the
+instance for `Ix (Fin 3)`.
+
+'### Underscore: Infer an explicit argument implicitly
+
+'Sometimes, a required function argument can be successfully inferred.
+In these cases one can write `_`, which means "fill this in by
+unification" (the same as implicit parameters).  For example,
+`from_ordinal` accepts the type to cast the ordinal to as an explicit
+argument, but when the index set of an array is obvious, there is no
+need to spell it:
+
+["foo", "bar"].(from_ordinal _ 0)
+> "foo"
+
+'Sadly, the reverse (explicitly supplying an implicit argument) is
+currently not possible (see [Issue
+464](https://github.com/google-research/dex-lang/issues/464)).  This
+is actually why `from_ordinal` takes the type as an explicit argument,
+so that it _can_ be supplied when it would not be inferrable.
+
+from_ordinal (Fin 2) 0
+> (0@Fin 2)
+
+'## Standalone function types
+
+'The syntax for spelling bare function types is slightly different
+from the syntax for defining functions with their type signatures.  To
+wit, for now (but see [Issue
+925](https://github.com/google-research/dex-lang/issues/925)), a
+function type uses different kinds of arrows to disambiguate different
+kinds of parameters.
+- The `->` arrow means "explicit parameter",
+- The `?->` arrow means "implicit parameter", and
+- The `?=>` arrow means "interface constraint".
+
+'(As usual, `=>` means "array type".)
+
+'For instance, the type of our `average` function above is
+
+:t average
+> ((n:Type) ?-> (v#0:(Ix n)) ?=> (n => Float32) -> Float32)
+
+'Confusingly ([Issue
+236](https://github.com/google-research/dex-lang/issues/236)), parens
+are not allowed around binders in standalone function types, so
+inputting that type into Dex would look like this:
+
+n:Type ?-> Ix n ?=> n => Float -> Float
+> (n:Type) ?-> (v#0:(Ix n)) ?=> (n => Float32) -> Float32
+
+'The effect specifier, as usual, is written in curly braces preceding the result type.
+
+'## Anonymous functions (lambda)
+
+'The syntax for anonymous functions is
+```
+\ <parameters> . <body>
+```
+The `body` can be a block.
+
+'The syntax for parameters is slightly different than allowed in
+`def`.  First, the `\` syntax permits bare explicit parameters
+(without parens), whose types are inferred.
+
+map (\a. a + 1) [1., 2., 3.]
+> [2., 3., 4.]
+
+'We can also explicitly type-annotate the arguments if we want, by wrapping
+them in parens like `def`.
+
+map (\(a:Float). a + 1) [1., 2., 3.]
+> [2., 3., 4.]
+
+'Like `def`, we can also specify implicit parameters with curly braces,
+
+\{a} (x:a). x
+> \a:Type ?-> \x:a. x
+
+'which default to type `Type` if not annotated.  However, the `\`
+syntax does not allow listing interface constraints, because that
+conflicts with pattern-matching a table argument (see [Issue
+954](https://github.com/google-research/dex-lang/issues/954)).
+
+\{n} [Ix n] (xs:n=>Float). sum xs
+> Type error:Not a data constructor: Ix
```

Comment:
On the contrary, I think we should, so that a hapless user can have some hope of knowing what it means.  If and when we fix #954, we can revisit this paragraph.

I did change it to mention that the error message is not very good, so a reader of the documentation does not try to infer anything from the assumption that error message is good.